### PR TITLE
[7/10] Code generation for Pooling and Fully Connected via CMSIS-NN

### PR DESCRIPTION
**Pull request by ashutosh-arm**
Support for Maxpool and AvgPool via CMSIS-NN.
Code generation for fully connected layer via CMSIS-NN

This PR depends on #9456.

Tracking issue #8646.

### CODE REVIEW

**Comment by ashutosh-arm** [comment]
@manupa-arm I think we should merge this one as this contains both of those changes once I've finally pushed changes addressing your comments.

**Comment by manupa-arm** [comment]
If we want to do that please close the other PR and adjust titles and descriptions of this PR to reflect its contents.

**Comment by ashutosh-arm** [comment]
> If we want to do that please close the other PR and adjust titles and descriptions of this PR to reflect its contents.

Done

File: src/relay/backend/contrib/cmsisnn/relay_to_tir.cc [hunk 1]
```diff
@@ -164,21 +156,15 @@ class RelayToTIRVisitor : public MixedModeMutator {
                                         ToArg(dilation_w),   ToArg(dilation_h),    ToArg(clip_min),
                                         ToArg(clip_max)};
 
-    // cmsis_nn_dims *input_dims (NHWC)
+    // layout NHWC
```
Comment (manupa-arm) [inline src/relay/backend/contrib/cmsisnn/relay_to_tir.cc:159]: Sorry Ashutosh, I dont still follow what this comment mean. How is this related to the following line ?

File: src/relay/backend/contrib/cmsisnn/relay_to_tir.cc [hunk 2]
```diff
@@ -164,21 +156,15 @@ class RelayToTIRVisitor : public MixedModeMutator {
...
     Array<PrimExpr> input_shape = conv2d_call->args[0]->type_as<TensorTypeNode>()->shape;
-    Array<PrimExpr> input_dims = CMSISNNDimensions(input_shape);
 
-    // cmsis_nn_dims *filter_dims (OHWI for Conv2D and IHWO for depthwise)
+    // OHWI for Conv2D and IHWO for depthwise
```
Comment (manupa-arm) [inline src/relay/backend/contrib/cmsisnn/relay_to_tir.cc:162]: Same here, where is the information OHWI is used ?

File: src/relay/backend/contrib/cmsisnn/relay_to_tir.cc [hunk 3]
```diff
@@ -194,7 +180,7 @@ class RelayToTIRVisitor : public MixedModeMutator {
     if (depth_multiplier != -1) {
       cmsisnn_api = "arm_depthwise_conv_wrapper_s8";
       Array<PrimExpr> depthwise_filter_shape{1, filter_shape[0], filter_shape[1], out_channels};
```
Comment (manupa-arm) [inline src/relay/backend/contrib/cmsisnn/relay_to_tir.cc:182]: Is it possible to use something like 
`
int kernel_pos_o = kernel_layout.find("O");
`
instead of the numbers ?

File: src/relay/backend/contrib/cmsisnn/relay_to_tir.cc [hunk 4]
```diff
@@ -234,6 +220,197 @@ class RelayToTIRVisitor : public MixedModeMutator {
...
+    }
+    call_ext_args.push_back(output);
+
+    int context_buffer_size = 0;
+    std::string context_buffer_name = "NULL";
+    tvm::Array<PrimExpr> context_buffer_args = {tir::StringImm(context_buffer_name),
+                                                ToArg(context_buffer_size)};
+
+    scalar_args = tvm::runtime::Concat(context_buffer_args, scalar_args);
+    scalar_args = tvm::runtime::Concat(scalar_args, cmsisnn_input_shape);
```
Comment (manupa-arm) [inline src/relay/backend/contrib/cmsisnn/relay_to_tir.cc:311]: [Clarity] I think we can bring the declaration "cmsisnn_input_shape"  closer to here as it is not used before this

File: src/relay/backend/contrib/cmsisnn/relay_to_tir.cc [hunk 5]
```diff
@@ -234,6 +220,197 @@ class RelayToTIRVisitor : public MixedModeMutator {
...
+    scalar_args = tvm::runtime::Concat(scalar_args, cmsisnn_filter_shape);
```
Comment (manupa-arm) [inline src/relay/backend/contrib/cmsisnn/relay_to_tir.cc:312]: [Clarity] I think we can bring the declaration "cmsisnn_filter_shape" closer to here as it is not used before this

File: src/relay/backend/contrib/cmsisnn/relay_to_tir.cc [hunk 6]
```diff
@@ -234,6 +220,197 @@ class RelayToTIRVisitor : public MixedModeMutator {
...
+    scalar_args = tvm::runtime::Concat(scalar_args, bias_shape);
```
Comment (manupa-arm) [inline src/relay/backend/contrib/cmsisnn/relay_to_tir.cc:313]: [Clarity] I think we can bring the declaration "bias_shape" closer to here as it is not used before this

File: src/relay/backend/contrib/cmsisnn/relay_to_tir.cc [hunk 7]
```diff
@@ -234,6 +220,197 @@ class RelayToTIRVisitor : public MixedModeMutator {
...
+    scalar_args = tvm::runtime::Concat(scalar_args, cmsisnn_output_shape);
```
Comment (manupa-arm) [inline src/relay/backend/contrib/cmsisnn/relay_to_tir.cc:314]: [Clarity] I think we can bring the declaration "cmsisnn_output_shape" closer to here as it is not used before this

**Review by manupa-arm — APPROVED** [review]
Broadly looks good! There are few minor comments related to clarity which I dont want to block the PR for another CI cycle. Would you be able to address them in a follow up ?

**Comment by ashutosh-arm** [comment]
Thanks for the reviews. 
Sure @manupa-arm, I can create a follow up for this. Please feel free to merge.

File: src/relay/backend/contrib/cmsisnn/relay_to_tir.cc (hunk 1 above)
Comment (ashutosh-arm) [reply src/relay/backend/contrib/cmsisnn/relay_to_tir.cc:159]: This was to help understand that the following shape is in NHWC layout which will be passed on to CMSIS-NN API eventually.

File: src/relay/backend/contrib/cmsisnn/relay_to_tir.cc (hunk 2 above)
Comment (ashutosh-arm) [reply src/relay/backend/contrib/cmsisnn/relay_to_tir.cc:162]: Same as above. Maybe I should explicitly say that this is for CMSIS-NN API?

File: src/relay/backend/contrib/cmsisnn/relay_to_tir.cc (hunk 4 above)
Comment (ashutosh-arm) [reply src/relay/backend/contrib/cmsisnn/relay_to_tir.cc:311]: We cannot. Its being used to derive another size in the middle. Also, I would prefer to keep concatenation of arguments at one place. Locating all arguments at same place has helped me with debugs.

File: src/relay/backend/contrib/cmsisnn/relay_to_tir.cc (hunk 5 above)
Comment (ashutosh-arm) [reply src/relay/backend/contrib/cmsisnn/relay_to_tir.cc:312]: Same as above.

File: src/relay/backend/contrib/cmsisnn/relay_to_tir.cc (hunk 6 above)
Comment (ashutosh-arm) [reply src/relay/backend/contrib/cmsisnn/relay_to_tir.cc:313]: Same as above.

File: src/relay/backend/contrib/cmsisnn/relay_to_tir.cc (hunk 7 above)
Comment (ashutosh-arm) [reply src/relay/backend/contrib/cmsisnn/relay_to_tir.cc:314]: Same as above.

File: src/relay/backend/contrib/cmsisnn/relay_to_tir.cc (hunk 1 above)
Comment (manupa-arm) [reply src/relay/backend/contrib/cmsisnn/relay_to_tir.cc:159]: I think might worth saying that explicitly :)

File: src/relay/backend/contrib/cmsisnn/relay_to_tir.cc (hunk 2 above)
Comment (manupa-arm) [reply src/relay/backend/contrib/cmsisnn/relay_to_tir.cc:162]: I think the information is that the following line produces a tensor is that layout and subsequent access assume that ?

File: src/relay/backend/contrib/cmsisnn/relay_to_tir.cc (hunk 1 above)
Comment (ashutosh-arm) [reply src/relay/backend/contrib/cmsisnn/relay_to_tir.cc:159]: Ack

File: src/relay/backend/contrib/cmsisnn/relay_to_tir.cc (hunk 4 above)
Comment (manupa-arm) [reply src/relay/backend/contrib/cmsisnn/relay_to_tir.cc:311]: I think Im missing something here :). Isn't it defined in line 291 and is there a use in between ?
Anyway, if it helps debugging, lets have it this way.

File: src/relay/backend/contrib/cmsisnn/relay_to_tir.cc (hunk 2 above)
Comment (ashutosh-arm) [reply src/relay/backend/contrib/cmsisnn/relay_to_tir.cc:162]: Yes, comment got displaced in the process of review and new edits :laughing:

**Comment by ashutosh-arm** [comment]
@manupa-arm Could you please put commit message as "Support for code generation of Maxpool, AvgPool and Fully Connected layers via CMSIS-NN"

File: src/relay/backend/contrib/cmsisnn/relay_to_tir.cc (hunk 4 above)
Comment (ashutosh-arm) [reply src/relay/backend/contrib/cmsisnn/relay_to_tir.cc:311]: Sorry, I was looking at a different operator. For some of the others, context buffer creation requires this shape. Just to be consistent, I didn't want to change the code layout.

**Comment by manupa-arm** [comment]
Thanks @ashutosh-arm ! This is merged now.
